File: tests/functional/Bots/JavaScript/Skills/CodeFirst/WaterfallSkillBot/dialogs/delete/deleteDialog.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { MessageFactory } = require('botbuilder');
const { ComponentDialog, WaterfallDialog, DialogTurnStatus } = require('botbuilder-dialogs');
const { Channels } = require('botbuilder-core');

const SLEEP_TIMER = 5000;
const WATERFALL_DIALOG = 'WaterfallDialog';
const DELETE_SUPPORTED = new Set([Channels.Msteams, Channels.Slack, Channels.Telegram]);

class DeleteDialog extends ComponentDialog {
  /**
   * @param {string} dialogId
   */
  constructor (dialogId) {
    super(dialogId);

    this.addDialog(new WaterfallDialog(WATERFALL_DIALOG, [
      this.handleDeleteDialog.bind(this)
    ]));

    this.initialDialogId = WATERFALL_DIALOG;
  }

  /**
   * @param {import('botbuilder-dialogs').WaterfallStepContext} stepContext
   */
  async handleDeleteDialog (stepContext) {
    const channel = stepContext.context.activity.channelId;

    if (DeleteDialog.isDeleteSupported(channel)) {
      const id = await stepContext.context.sendActivity(MessageFactory.text('I will delete this message in 5 seconds'));
      await DeleteDialog.sleep(SLEEP_TIMER);
      await stepContext.context.deleteActivity(id.id);
    } else {
      await stepContext.context.sendActivity(MessageFactory.text(`Delete is not supported in the ${channel} channel.`));
    }

    return { status: DialogTurnStatus.complete };
  }

  /**
   * @param {number} milliseconds
   */
  static sleep (milliseconds) {
    return new Promise(resolve => {
      setTimeout(resolve, milliseconds);
    });
  }

  /**
   * @param {string} channel
   */
  static isDeleteSupported (channel) {
    return DELETE_SUPPORTED.has(channel);
  }
}

module.exports.DeleteDialog = DeleteDialog;

File: tests/functional/Bots/Python/Skills/CodeFirst/WaterfallSkillBot/dialogs/delete/delete_dialog.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

import time
from botbuilder.core import MessageFactory
from botbuilder.dialogs import (
    ComponentDialog,
    WaterfallDialog,
    WaterfallStepContext,
    DialogTurnResult,
    DialogTurnStatus,
)
from botframework.connector import Channels


class DeleteDialog(ComponentDialog):
    def __init__(self):
        super().__init__(DeleteDialog.__name__)
        self._delete_supported = [Channels.ms_teams, Channels.slack, Channels.telegram]

        self.add_dialog(
            WaterfallDialog(WaterfallDialog.__name__, [self.handle_delete_dialog_step])
        )

        self.initial_dialog_id = WaterfallDialog.__name__

    async def handle_delete_dialog_step(self, step_context: WaterfallStepContext):
        channel = step_context.context.activity.channel_id

        if channel in self._delete_supported:
            response = await step_context.context.send_activity(
                MessageFactory.text("I will delete this message in 5 seconds")
            )
            time.sleep(5)
            await step_context.context.delete_activity(response.id)

        else:
            await step_context.context.send_activity(
                MessageFactory.text(
                    f"Delete is not supported in the {channel} channel."
                )
            )

        return DialogTurnResult(DialogTurnStatus.Complete)
